cpp/ql/lib/semmle/code/cpp/dataflow/internal/FlowSummaryImpl.qll

```diff
@@ -54,42 +54,55 @@ module Input implements InputSig<DataFlowImplSpecific::CppDataFlow> {
   bindingset[token]
   ParameterPosition decodeUnknownParameterPosition(AccessPath::AccessPathTokenBase token) {
     // needed to support `Argument[x..y]` ranges, `Argument[-1]`, and indirections `*Argument[0]`.
-    exists(int indirection |
-      token.getName() = indirectionString(indirection) + "Argument" and
-      exists(int pos | pos = AccessPath::parseInt(token.getAnArgument()) |
-        pos >= 0 and indirection = 0 and result = TDirectPosition(pos)
+    exists(int indirection, string argPosString, int argPos |
+      token.getName() = "Argument" and
+      token.getAnArgument() = indirectionString(indirection) + argPosString and
+      argPos = AccessPath::parseInt(argPosString) and
+      (
+        argPos >= 0 and indirection = 0 and result = TDirectPosition(argPos)
         or
-        pos >= 0 and indirection > 0 and result = TIndirectionPosition(pos, indirection)
+        argPos >= 0 and indirection > 0 and result = TIndirectionPosition(argPos, indirection)
         or
         // `Argument[-1]` is the qualifier object `*this`, not the `this` pointer itself
-        pos = -1 and result = TIndirectionPosition(pos, indirection + 1)
+        argPos = -1 and result = TIndirectionPosition(argPos, indirection + 1)
       )
     )
   }
 
   bindingset[token]
   ArgumentPosition decodeUnknownArgumentPosition(AccessPath::AccessPathTokenBase token) {
     // needed to support `Argument[x..y]` ranges, `Argument[-1]`, and indirections `*Argument[0]`.
-    exists(int indirection |
-      token.getName() = indirectionString(indirection) + "Parameter" and
-      exists(int pos | pos = AccessPath::parseInt(token.getAnArgument()) |
-        pos >= 0 and indirection = 0 and result = TDirectPosition(pos)
+    exists(int indirection, string paramPosString, int paramPos |
+      token.getName() = "Parameter" and
+      token.getAnArgument() = indirectionString(indirection) + paramPosString and
+      paramPos = AccessPath::parseInt(paramPosString) and
+      (
+        paramPos >= 0 and indirection = 0 and result = TDirectPosition(paramPos)
         or
-        pos >= 0 and indirection > 0 and result = TIndirectionPosition(pos, indirection)
+        paramPos >= 0 and indirection > 0 and result = TIndirectionPosition(paramPos, indirection)
         or
         // `Argument[-1]` is the qualifier object `*this`, not the `this` pointer itself
-        pos = -1 and result = TIndirectionPosition(pos, indirection + 1)
+        paramPos = -1 and result = TIndirectionPosition(paramPos, indirection + 1)
       )
     )
   }
 
   bindingset[token]
   ContentSet decodeUnknownContent(AccessPath::AccessPathTokenBase token) {
-    // field content (with indirection support).
+    // field content (no indirection support)
     exists(FieldContent c |
       result.isSingleton(c) and
+      token.getName() = c.getField().getName() and
+      not exists(token.getArgumentList()) and
+      c.getIndirectionIndex() = 1
```

Comment:
We need to be careful about the join ordering here: if we join with `c.getField()` first we can get an explosion by a factor of max-number-of-indirections-for-a-field-in-the-DB (probably 4 or 5, I'd imagine), and if we join with `c.getIndirectionIndex()` first we can get an explosion by a factor of number-of-fields-in-the-DB (thousands or millions).

It may be that the compiler does the right thing here (i.e., turn this into a join on both columns) automatically, but if it doesn't we may need to slightly rewrite this (and the case below as well).

Reply:
There are thousands of similar cases throughout our CodeQL and IMO the only reasonable way to deal with them is to trust the optimizer to do the right thing (at least until there's evidence it has failed).  What worries you about this case in particular?  Is there a better way I could write it without scattering `pragma`s around prematurely?

Reply:
I don't think we should scatter `pragma`s around prematurely. That's why I said:
> It may be that the compiler does the right thing here (i.e., turn this into a join on both columns) automatically, but if it doesn't we may need to slightly rewrite this (and the case below as well).

Reading what I wrote now, I do see that it wasn't very clear. Sorry! What I meant to say was that we should go back and revisit this code if it turns out we end up seeing a join order issue here.

My spider sense was tingling because we've seen quite a lot of bad joins caused by code such as this in the past:
```ql
indirectOperand.getOperand() = op and
indirectOperand.getIndirectionIndex() = i and
somehowRelatedOperandAndIndex(op, i)
```
and the fix has been to rewrite this to look like:
```ql
indirectOperand.hasOperandAndIndirectionIndex(op, i) // I forgot what the actual predicate is called, but you get the point
somehowRelatedOperandAndIndex(op, i)
```

Reply:
> (at least until there's evidence it has failed)

Totally agree

Reply:
OK, so in this case the fix (if we want it) would be to change:
```
token.getName() = c.getField().getName() and
...
c.getIndirectionIndex() = 1
```
into something like:
```
token.getName() = c.getFieldWithIndirectionIndex(1).getName() and
```

Reply:
Yeah, something like that. But let's save that until we actually see a problem, if any.

Reply:
In saw no performance problems on the (first) DCA run, and I'm not seeing `decodeUnknownContent` come up in as an expensive predicate in performance logs ... so 🤞 this is indeed not actually a problem.